gwu-libraries/scholarspace-hyrax
# T391 1 dev test docker

# PR 1 of 3 for being able to run automated test and CI/CD (#391) 

Tried to separate out just the code changes related to Docker and Rails environments that are needed for running RSpec tests and running in development mode in Docker, so the immediate benefits of these changes may not be evident. After testing of this PR, next are changes required for setting up RSpec and other testing gems, followed by the RSpec tests. 

The biggest changes are: 
- The naming of the Hyrax databases, Fedora databases, and Solr cores. These have been refactored to just be set with a single `.env` variable each, and automatically appends `_dev` and `_test` in configurations for consistency. 
- Changing default bind mounts for backup Fedora and Solr data on local machine.
- Changing the Rails default/dev server to use Passenger rather than Puma
- Fixing `json-canonicalization` gem error
- Exposing port 3000 in `docker-compose` so we can use Rails development server
- Rather than using templates to map files in/out of docker container, use non-template versions - easier when developing in development mode in docker container. 

# To test: 
- Build and start docker containers:
  - `docker-compose build`
  - `docker-compose up -d`
  - (or however else you manage docker containers, e.g. VSCode extension)
- Follow README instructions to set up in `production` mode, either following instructions for restoring from backup data, or creating a new production instance (e.g. `docker exec -it --user scholarspace [app-server-container-name] bash -lc "rails db:{drop,create,migrate,seed}"`)
- Follow the instructions in the "Development Mode" portion of the README to prepare, launch, and stop a development instance accessible via port 3000. If successful, you should be able to interact with the development instance and the production instance simultaneously in different browser tabs.
